staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -385,28 +385,60 @@ func (*Mounter) List() ([]MountPoint, error) {
 	return ListProcMounts(procMountsPath)
 }
 
+// statx support since Linux 4.11, glibc 2.28
+// refer: https://man7.org/linux/man-pages/man2/statx.2.html
+var errNotSupport = errors.New("The statx syscall is not supported. At least Linux kernel 4.11 is needed")
+
+func statx(file string) (unix.Statx_t, error) {
+	// TODO add STATX_ATTR_MOUNT_ROOT support
+	// Linux 5.8 commit 80340fe3605c0e78cfe496c3b3878be828cfdbfe
+	// stat->attributes |= STATX_ATTR_MOUNT_ROOT;
+	// stat->attributes_mask |= STATX_ATTR_MOUNT_ROOT;
+
+	var stat unix.Statx_t
+	if err := unix.Statx(0, file, unix.AT_STATX_DONT_SYNC, 0, &stat); err != nil {
+		if err == unix.ENOSYS {
+			return stat, errNotSupport
+		}
+
+		return stat, err
+	}
+
+	return stat, nil
+}
+
 // IsLikelyNotMountPoint determines if a directory is not a mountpoint.
-// It is fast but not necessarily ALWAYS correct. If the path is in fact
-// a bind mount from one part of a mount to another it will not be detected.
-// It also can not distinguish between mountpoints and symbolic links.
+// If fast check failed, fall back to slow path. If the path is in fact
+// a bind mount from one part of a mount to another it will be detected.
+// It also can distinguish between mountpoints and symbolic links.
 // mkdir /tmp/a /tmp/b; mount --bind /tmp/a /tmp/b; IsLikelyNotMountPoint("/tmp/b")
-// will return true. When in fact /tmp/b is a mount point. If this situation
-// is of interest to you, don't use this function...
+// will return false. When in fact /tmp/b is a mount point.
+
+// TODO(j4ckstraw) add test
 func (mounter *Mounter) IsLikelyNotMountPoint(file string) (bool, error) {
-	stat, err := os.Stat(file)
-	if err != nil {
-		return true, err
-	}
-	rootStat, err := os.Stat(filepath.Dir(strings.TrimSuffix(file, "/")))
-	if err != nil {
-		return true, err
+	var stat, rootStat unix.Statx_t
+	var err error
+
+	if stat, err = statx(file); err != nil {
```

Comment:
Would it be better to reuse [MountFast](https://github.com/kubernetes/kubernetes/blob/a2c46721638f9d9d33eadbf596df0bfa7a543ca9/vendor/github.com/moby/sys/mountinfo/mounted_linux.go#L24) that needs 5.10 ?

It falls back to `stat()`, but only in < 5.10

Reply:
I'm afraid not, because [normalizePath](https://github.com/kubernetes/kubernetes/blob/a2c46721638f9d9d33eadbf596df0bfa7a543ca9/vendor/github.com/moby/sys/mountinfo/mounted_unix.go#L40) will call Stat again.

Reply:
rather than falling back to `IsMountPoint` - should it fall back to a implementation of `IsLikelyNotMountPoint` which uses `stat` ?

Reply:
Sorry, I don't understand what you mean.

The key point here is that `stat` syscall will trap in kernel, which make kubelet struck in main loop. So I replace `stat` with `statx`.  AT_STATX_DONT_SYNC option make statx don't sync with remote nfs server, so will not trap in kernel.

English is not my native language. Please forgive me if there are any grammatical mistakes :)

Reply:
Good catch. IMO it would make sense to use `statx` in moby's `normalizePath` then
@kolyshkin, WDYT?

Reply:
@j4ckstraw I was saying on kernels that don't support `statx` syscalls, rather than going to `IsMountPoint` which is slow and could be considered a regression in certain environments, we fall back to an implementation of `IsLikelyNotMountPoint` which still uses `stat` and yes, it will get kubelet stuck but it _at least_ behave same as older version and will not be considered a regression.

Reply:
@gnufied Updated, it's fallback to origin `IsLikelyNotMountPoint` now.
As you can see, I change the comment of `IsLikelyNotMountPoint`, I change the semantic of the `IsLikelyNotMountPoint`in the last patch, so it is reliable, but with risk of regression and slow if the kernel not support `statx`.

After updated, it may be free of regression, but still have risk of making kubelet struck in main loop.
WDYT?